### PR TITLE
Assorted fixes

### PR DESCRIPTION
**Pull request by Ralith**
In particular, delaying handshake key discard reduces odds of connection loss under severe packet loss following connection establishment.

### CODE REVIEW

File: quinn-proto/src/shared.rs [hunk 1]
```diff
@@ -311,3 +311,31 @@ pub struct ClientOpts {
...
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
+pub struct ResetToken([u8; RESET_TOKEN_SIZE]);
+
+impl From<[u8; RESET_TOKEN_SIZE]> for ResetToken {
+    fn from(x: [u8; RESET_TOKEN_SIZE]) -> Self {
+        Self(x)
+    }
+}
+
+impl std::ops::Deref for ResetToken {
```
Comment (djc) [inline quinn-proto/src/shared.rs]: Could we `Deref` in a way that we don't need `[..]` in a few places?

File: quinn-proto/src/connection.rs [hunk 2]
```diff
@@ -541,7 +541,9 @@ impl Connection {
 
     fn set_key_discard_timer(&mut self, now: Instant) {
         let time = if self.spaces[SpaceId::Handshake as usize].crypto.is_some() {
-            now + self.pto() * 3
+            // Severe packet loss during this period can cause the connection to fail, so allocate
```
Comment (djc) [inline quinn-proto/src/connection.rs]: Maybe add something to the comment about the potential downside of what we're trading off here?

File: quinn-proto/src/shared.rs (hunk 1 above)
Comment (Ralith) [reply quinn-proto/src/shared.rs]: I tweaked the `extend` call; can't do much for `==` without adding a significantly larger boilerplatey manual `Eq` impl, I think.

File: quinn-proto/src/connection.rs (hunk 2 above)
Comment (Ralith) [reply quinn-proto/src/connection.rs]: Elaborated to the best of my understanding; might be worth double-checking on slack or against TLS references, though.

Comment (Ralith) [reply quinn-proto/src/connection.rs]: The TLS draft [is very relaxed](https://quicwg.org/base-drafts/draft-ietf-quic-tls.html#discarding-unused-keys) about requirements here; handshake key discard does not appear to actually be mandated at all, though I understand a future draft may be more specific about reasonable behavior here.

**Pull request merged**